bot/cogs/regular_expressions.py

```diff
@@ -0,0 +1,126 @@
+import re
+from typing import Iterable, Optional, Union
+
+import regex
+from discord.ext.commands import BadArgument, Cog, Context, Converter, group
+
+from bot.bot import Bot
+
+
+def format_error(e: Union[re.error, regex.error]) -> Iterable[str]:
+    r"""
+    Format a regexp parsing error message to display in a response.
+
+    >>> try: re.compile("\w+(")
+    ... except re.error as e: print(format_error(e))
+    ['\w+(', '   ^', 'missing ), unterminated subpattern']
+
+    Which will look like:
+    | \w+(
+    |    ^
+    | missing ), unterminated subpattern
+    """
+    line_with_regexp = e.pattern
+    line_with_caret = " "*e.pos + "^"
+    line_with_explanation = e.msg
+    return [line_with_regexp, line_with_caret, line_with_explanation]
+
+
+def format_match(match: Optional[re.Match]) -> Iterable[str]:
+    r"""
+    Format a match result to display in a response.
+
+    >>> format_match(re.search("(\d\d)+", "hello123456world"))
+    ['    hello123456world', ' 0:      ^^^^^^', ' 1:          ^^']
+
+    Which will look as:
+    |    hello123456world
+    | 0:      ^^^^^^
+    | 1:          ^^
+    """
```

Comment:
What do you think about showing the actual text matched instead of the carets? That may end up being easier to read, especially if there are a lot of groups (eyes won't have to travel far to match the carets to the original string).

Reply:
I think this would be better:
```
    |    hello123456world
    | 0:      123456
    | 1:          56
```
just showing the text might be ambiguous if a substring appears multiple times in a string.

However, we still have to figure out how to display zero-width matches :thinking: